### ISSUE TITLE
False positive match_wildcard_for_single_variants on enum with hidden variant

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#6984]
Beginning with the most recent nightly (clippy 0.1.53 5e65467 2021-03-26, rustc 1.53.0-nightly 5e65467ef 2021-03-26) the following code incorrectly triggers `match_wildcard_for_single_variants`. I believe clippy should not be recommending that the user write doc(hidden) variants into their code.

```toml
# .clippy.toml

msrv = "1.39.0"
```

```rust
// src/lib.rs

pub enum Enum {
    A,
    B,
    #[doc(hidden)]
    __Private,
}
```

```rust
// src/main.rs

#![warn(clippy::pedantic)]

use testing::Enum;

fn main() {
    let e = Enum::A;
    let _ = match e {
        Enum::A => 'a',
        Enum::B => 'b',
        _ => 'c',
    };
}
```

```console
warning: wildcard matches only a single variant and will also match any future added variants
  --> src/main.rs:10:9
   |
10 |         _ => 'c',
   |         ^ help: try this: `Enum::__Private`
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::match_wildcard_for_single_variants)]` implied by `#[warn(clippy::pedantic)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_wildcard_for_single_variants
```

Mentioning @Jarcho @llogiq who touched this lint recently in #6863.
